### PR TITLE
Enable vision bucketing/warmup for gemma3 model

### PR DESCRIPTION
**Pull request by libinta**
Enable Gemma3 vl vision_tower bucketing with VLLM_MULTIMODAL_BUCKETS:
1. define vision bucket based on batch, (default to [1,2,4,8])
2.  combine all images in a request to get image_features from _image_pixels_to_features
3.  split them to batches based on the predefined buckets to get vision embedding through multi_modal_projector
5.  get inputs_embeds from  vision_embedding+inputs_ids through get_input_embeddings
6.  pass the inputs_embeds to language model to get hidden states

### CODE REVIEW

File: vllm/attention/backends/hpu_attn.py [hunk 1]
```diff
@@ -397,6 +403,7 @@ def __init__(
             else ModuleFusedSDPA(HPUFusedSDPA)
 
         self.prefill_impl = get_config().prompt_attn_impl
+
```
Comment (michalkuligowski) [inline vllm/attention/backends/hpu_attn.py:406]: new line, please remove

File: vllm/attention/backends/hpu_attn.py [hunk 2]
```diff
@@ -554,9 +563,21 @@ def forward(
                 **self.common_attention_args(block_list, key_cache,
                                              value_cache,
                                              attn_metadata.block_size))
+
             output = out.reshape(batch_size, seq_len, hidden_size)
         else:
             # Decoding run.
+            if not self.sliding_window:
```
Comment (michalkuligowski) [inline vllm/attention/backends/hpu_attn.py:570]: please revers condition, its easier to read when its:
if self,sliding_window:

File: vllm/model_executor/models/gemma3_mm.py [hunk 3]
```diff
@@ -541,32 +546,56 @@ def _parse_and_validate_image_input(
...
 
     def _image_pixels_to_features(
-        self,
-        vision_tower: SiglipVisionModel,
-        pixel_values: torch.Tensor,
-    ) -> torch.Tensor:
+            self,
+            vision_tower: SiglipVisionModel,
+            pixel_values: torch.Tensor,
+            graphed_multimodal_buckets=None) -> torch.Tensor:
```
Comment (michalkuligowski) [inline vllm/model_executor/models/gemma3_mm.py:560]: you are modifing model code, which is shared across vendors, please use custom hpu version

File: vllm/model_executor/models/gemma3_mm.py [hunk 4]
```diff
@@ -541,32 +546,56 @@ def _parse_and_validate_image_input(
...
         image_features = vision_tower(pixel_values.to(dtype=target_dtype))
         return image_features
 
     def _process_image_input(
-        self,
-        image_input: Gemma3ImageInputs,
-    ) -> list[torch.Tensor]:
+            self,
+            image_input: Gemma3ImageInputs,
+            graphed_multimodal_buckets=None) -> list[torch.Tensor]:
```
Comment (michalkuligowski) [inline vllm/model_executor/models/gemma3_mm.py:568]: as above

File: vllm/model_executor/models/gemma3_mm.py [hunk 5]
```diff
@@ -639,30 +672,42 @@ def prepare_attn_masks(
...
+            start_idices = (positions == 0).cpu().nonzero()
+            num_seqs = len(start_idices)
+            for i in range(num_seqs):
+                start_idx = start_idices[i].item()
+                if i < num_seqs - 1:
+                    end_idx = start_idices[i + 1].item()
+                else:
+                    end_idx = len(input_ids)
+                seq_lens.append(end_idx - start_idx)
+                kwargs["seq_lens"] = seq_lens
```
Comment (michalkuligowski) [inline vllm/model_executor/models/gemma3_mm.py:695]: This doesn't work, intendation for kwargs["seq_lens"] is incorrect

File: vllm/model_executor/models/gemma3_mm.py [hunk 6]
```diff
@@ -639,30 +672,42 @@ def prepare_attn_masks(
...
 
         global_attn_masks = []
         local_attn_masks = []
         start_idx = 0
         for seq_len in seq_lens:
-            end_idx = start_idx + seq_len
-            input_token_ids = input_ids[start_idx:end_idx]
-            start_idx = end_idx
+            if not is_hpu:
```
Comment (michalkuligowski) [inline vllm/model_executor/models/gemma3_mm.py:701]: Please reverse the if-else clause

File: vllm/model_executor/models/gemma3_mm.py [hunk 7]
```diff
@@ -676,11 +721,32 @@ def prepare_attn_masks(
             # Consider the bidirectional attention between image tokens.
             img_mask = torch.zeros_like(global_attn_mask)
             img_pos = (input_token_ids == self.config.image_token_index)
-            img_mask[:, :, :, img_pos] += 1
-            img_mask[:, :, img_pos, :] += 1
-            global_attn_mask = torch.where(img_mask == 2, 0, global_attn_mask)
+            if not is_hpu:
```
Comment (michalkuligowski) [inline vllm/model_executor/models/gemma3_mm.py:724]: as above

**Review by michalkuligowski — CHANGES_REQUESTED** [review]

File: vllm/model_executor/models/gemma3_mm.py (hunk 6 above)
Comment (libinta) [reply vllm/model_executor/models/gemma3_mm.py:701]: done

File: vllm/model_executor/models/gemma3_mm.py (hunk 5 above)
Comment (libinta) [reply vllm/model_executor/models/gemma3_mm.py:697]: as it's from upstream, it works for non-hpu case

Comment (michalkuligowski) [reply vllm/model_executor/models/gemma3_mm.py:694]: Your PR has different indentation than the code you linked, it needs to be fixed before merging.

Comment (libinta) [reply vllm/model_executor/models/gemma3_mm.py]: DONE

**Comment by michalkuligowski** [comment]
need to run whole suite due to new fixes
/run-gaudi-tests

**Review by michalkuligowski — APPROVED** [review]
